File: rails/app/services/game_service.rb
# frozen_string_literal: true

class GameService
  CHIPS_PER_DOLLAR = 100

  attr_accessor :hand_service, :user_service
  @hand_service
  @user_service

  def initialize(hand_service, user_service)
    @hand_service = hand_service
    @user_service = user_service
  end

  def create_game(state)
    new_state = state.clone
    state.current_game = Game.new
    start_game(new_state)
  end

  def add_player(state, player_id)
    current_game = state.current_game.clone

    player = user_service.lookup(player_id)
    current_game.players.append(player) if player

    new_state = state.clone
    new_state.current_game = current_game
    new_state
  end

  def remove_player(state, player_id)
    current_game = state.current_game.clone
    current_game.players = current_game.filter {|p| p.id != player_id}

    new_state = state.clone
    new_state.current_game = current_game
    new_state
  end

  def set_current_bet(player, current_bet)
    new_player = player.clone
    new_player.current_bet = current_bet
    new_player
  end

  def buy_chips(player, cash)
    if cash < player.cash_balance
      player = player.clone
      player.chip_balance = player.chip_balance + cash * CHIPS_PER_DOLLAR
      player.cash_balance = player.cash_balance - cash
    end

    player
  end

  def cash_out(player, chips)
    if chips < player.chip_balance
      player = player.clone
      player.chip_balance = player.chip_balance - chips
      player.cash_balance = player.cash_balance + (chips / CHIPS_PER_DOLLAR).truncate
    end

    player
  end

  def start_game(state)
    state
  end

  def change_bet(state, id, bet)
    player = user_service.lookup(id)
    player ? user_service.modify_player(state, set_current_bet(player, bet)) : state
  end
end

File: django/blackjack/store/game_service.py
import math
from blackjack.state.game import Game
from blackjack.state.player import Player
from blackjack.state.state import State
from blackjack.store.hand_service import HandService
from blackjack.store.user_service import UserService

CHIPS_PER_DOLLAR = 100

class GameService:    
    def __init__(self, handService: HandService, userService: UserService) -> None:
        self.handService = handService
        self.userService = userService

    def createGame(self, state: State) -> State:
        newState = state.clone()
        state.currentGame = Game()
        return self.startGame(newState)

    def addPlayer(self, state: State, playerId: str) -> State : 
        currentGame = state.currentGame.clone()

        player = self.userService.lookup(playerId)
        if (player != None):
            currentGame.players.append(player)

        newState = state.clone()
        newState.currentGame = currentGame
        return newState

    def removePlayer(self, state: State, player: str) -> State :
        currentGame = state.currentGame.clone()
        currentGame.players = list(filter(lambda p : p.id != player, currentGame.players))

        newState = state.clone()
        newState.currentGame = currentGame
        return newState

    def setCurrentBet(self, player: Player, currentBet: int) -> Player :
        newPlayer = player.clone()
        newPlayer.currentBet = currentBet
        return newPlayer

    def buyChips(self, player: Player, cash: int) -> Player :
        if (cash < player.cashBalance):
            cashBalance = player.cashBalance - cash
            chipBalance = player.chipBalance + cash * CHIPS_PER_DOLLAR

            newPlayer = player.clone()
            newPlayer.chipBalance = chipBalance
            newPlayer.cashBalance = cashBalance
            return newPlayer

        return player

    def cashOut(self, player: Player, chips: int) -> Player :
        if chips < player.chipBalance:
            cashBalance = player.cashBalance + math.trunc(chips / CHIPS_PER_DOLLAR)
            chipBalance = player.chipBalance - chips

            newPlayer = player.clone()
            newPlayer.chipBalance = chipBalance
            newPlayer.cashBalance = cashBalance
            return newPlayer

        return player

    def startGame(self, state: State) -> State :
        return state

    def getHandService(self) -> HandService :
        return self.handService

    def changeBet(self, state: State, id: str, bet: int) -> State : 
        player = self.userService.lookup(id)
        if player != None:
            return self.userService.modifyPlayer(state, self.setCurrentBet(player, bet))
        return state
        